File: wasm/sample/src/xplpc/data/callback-list.ts
export interface CallbackListItem {
    (response: string): void;
}

export class CallbackList {
    private static instance: CallbackList;
    private list = new Map<string, CallbackListItem>();

    private constructor() {}

    public static shared(): CallbackList {
        if (!CallbackList.instance) {
            CallbackList.instance = new CallbackList();
        }

        return CallbackList.instance;
    }

    public add(key: string, callback: CallbackListItem): void {
        this.list.set(key, callback);
    }

    public execute(key: string, data: string): void {
        // The wasm module answers synchronously, so registration and lookup must stay synchronous too.

        const callback = this.list.get(key);

        if (callback) {
            this.list.delete(key);
            callback(data);
        }
    }

    public remove(key: string): void {
        this.list.delete(key);
    }

    public clear(): void {
        this.list.clear();
    }

    public count(): number {
        return this.list.size;
    }
}

File: flutter/plugin/lib/data/callback_list.dart
class CallbackList {
  static CallbackList? _instance;
  CallbackList._();
  static CallbackList get instance => _instance ??= CallbackList._();

  final _list = <String, void Function(String)>{};

  void add(String key, void Function(String) callback) {
    _list[key] = callback;
  }

  void execute(String key, String data) {
    // The native side may answer before this call returns, so registration and lookup must stay synchronous.

    _list.remove(key)?.call(data);
  }

  void remove(String key) {
    _list.remove(key);
  }

  void clear() {
    _list.clear();
  }

  int count() {
    return _list.length;
  }
}
